File: java_app/src/main/java/com/petnagy/java/example/redux/AppReducer.java
package com.petnagy.java.example.redux;

import com.petnagy.koredux.Action;
import com.petnagy.koredux.Reducer;

import org.jetbrains.annotations.NotNull;

public class AppReducer implements Reducer<AppState> {
    @Override
    public AppState invoke(@NotNull Action action, AppState appState) {
        AppState state = appState;
        if (action instanceof IncreaseAction) {
            state = new AppState(state.getCounter() + 1);
        } else if (action instanceof DecreaseAction) {
            state = new AppState(state.getCounter() - 1);
        }
        return state;
    }
}

File: app/src/main/java/com/petnagy/app/redux/AppReducer.kt
package com.petnagy.app.redux

import com.petnagy.koredux.Action
import com.petnagy.koredux.Reducer

class AppReducer: Reducer<AppState> {
    override fun invoke(action: Action, state: AppState): AppState {
        return state.copy(counter = counterReducer(action, state.counter))
    }

    private fun counterReducer(action: Action, actualCounter: Int): Int {
        return when(action) {
            is IncreaseAction -> actualCounter + 1
            is DecreaseAction -> actualCounter - 1
            else -> 0
        }
    }
}